botocore/endpoint.py

```diff
@@ -199,24 +199,24 @@ def _get_response(self, request, operation_model, attempts):
                                                  operation_model)
         parser = self._response_parser_factory.create_parser(
             operation_model.metadata['protocol'])
-        return ((http_response,
-                 parser.parse(response_dict, operation_model.output_shape)),
-                None)
+        parsed_response = parser.parse(
+            response_dict, operation_model.output_shape)
+        return (http_response, parsed_response), None
 
     def _looks_like_dns_error(self, e):
         return 'gaierror' in str(e) and e.request is not None
 
     def _looks_like_bad_status_line(self, e):
         return 'BadStatusLine' in str(e) and e.request is not None
 
-    def _needs_retry(self, attempts, operation_model, response=None,
-                     caught_exception=None):
+    def _needs_retry(self, attempts, operation_model, request_dict,
+                     response=None, caught_exception=None):
         event_name = 'needs-retry.%s.%s' % (self._endpoint_prefix,
                                             operation_model.name)
         responses = self._event_emitter.emit(
             event_name, response=response, endpoint=self,
             operation=operation_model, attempts=attempts,
-            caught_exception=caught_exception)
+            caught_exception=caught_exception, request_dict=request_dict)
```

Comment:
Instead of updating it [here](https://github.com/boto/botocore/pull/936/commits/f32d6241da8c137008e93a9e3647127a2a2a4c59#diff-cb00fc815f8f80ae15e7d0cecc74b395R311) you could just assign the reference to it and that should allow you to modify it. I do not see why it even needs to be an update since you start with a brand new dictionary each time. I honestly think you should even be able to inject it into the ``AWSRequest`` object. I am just really trying to avoid adding a new parameter to an event if we do not need to.